Add sensors listing and data streaming

llucax opened the pull request:
This pull request introduces support for sensors to the microgrid client.

A key focus of this update is **forward-compatibility** with the upcoming microgrid API v0.17. The new sensor-related RPCs (`list_sensors()`, `stream_sensor_data()`) and their associated data structures have been designed based on the v0.17 specifications. This approach aims to simplify the future migration process for users.

Fixes #136.

llucax committed:
Add a sensor ID wrapper

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Add a Lifetime wrapper

The lifetime will be introduced in the next microgrid API version, but
we add it now so we can add a rpc call to get sensors information in a
way that is compatible with future microgrid API versions.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Add a utility to convert protobuf enums to Python enums

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Add a `Sensor` class

This class will be used to list and retrieve sensors information from
the microgrid. The class is based on version 0.17 of the microgrid API
instead of the current 0.15 version, so users can already use the new
interface and make migrating to 0.17 more straight forward.

It is also implemented having in mind that sensor types will be removed
in future releases (and it is actually not being set properly in the
0.15 service), so sensor types are being ignored.

To be compatible with 0.17, sensors also have a "fake" lifetime that is
not really retrieved from the microgrid (it is filled with a "eternal"
lifetime, so sensors are present always).

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Add a `list_sensors()` RPC to get sensors

This RPC is also based on version 0.17 of the microgrid API instead of
the current 0.15 version, so users can already use the new interface and
make migrating to 0.17 more straight forward.

It uses the `ListComponents` RPC filtering to retrieve only sensors.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Add a metric aggregation wrapper

These wrappers are also not present in 0.15, but are added so users can
already start writing code that supports aggregates to ease the
migration to 0.17.

Because of this, no protobuf parsing code is added yet.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Add wrappers for sensor data

Again, these wrappers are done with 0.17 in mind, and trying to be
forward-compatible, although sensor data in 0.15 is already more similar
to 0.17 (compared to component data).

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Add a `stream_sensor_data()` RPC to get sensor data

This RPC is also based on version 0.17 of the microgrid API instead of
the current 0.15 version, so users can already use the new interface
and make migrating to 0.17 more straight forward.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Update release notes

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax commented:
@shsms I should have addressed all comments in #144. @tiyash-basu-frequenz maybe it would be good if you have a look to see if there is nothing that goes against future directions.

llucax commented:
OK, it looks like we are going to unify the `SensorMetric` and `Metric` for components and the metric samples too. But I think I would do that in a separate PR once it is more settled, as this is super fresh.

llucax commented:
BTW, now I tested this in a real microgrid and it seems to be working fine.

llucax commented:
@shsms I hope you are happier with the commit split now too :pray:

llucax commented:
Will rename `Lifetime.active` to `Lifetime.is_operational_now()` (no property, since it is an "expensive" operation), and `active_at` -> `is_operational_at()`.

shsms commented on src/frequenz/client/microgrid/_id.py:
```diff
@@ -105,3 +105,53 @@ def __repr__(self) -> str:
+
+    def __eq__(self, other: object) -> bool:
+        """Check if this instance is equal to another object."""
+        # This is not an unidiomatic typecheck, that's an odd name for the check.
```
I feel like whether it is odd or not is out of scope for this comment.

```suggestion
        # This is not an unidiomatic typecheck, pylint is wrong.

llucax replied:
Fair, not convinced about the plain **wrong** here, but I'll see if I can put it in a more relevant way. Although this is in every ID class, so I would have to update all comments.

shsms replied:
It says it is unidiomatic, you say it is not, so it is wrong.  But more context sounds good, of course.

llucax replied:
**THANK YOU**. I had to redo the whole module because of this :stuck_out_tongue_closed_eyes:

shsms commented on src/frequenz/client/microgrid/_util.py:
```diff
@@ -0,0 +1,50 @@
+    """
+    # A protobuf enum value is a NewType of int, so we make sure we have a pure int at
+    # this point.
+    value = int(value)
```
NewType is a static-checker-only thing.  At run time, it is actually just the int value, so this cast is not necessary.

From: https://docs.python.org/3/library/typing.html#newtype

> More precisely, the expression some_value is Derived(some_value) is always true at runtime.

shsms commented on tests/test_sensor.py:
```diff
@@ -0,0 +1,152 @@
+    model_name=_SENSOR.model_name,
+)
+
+_DIFFERENT_ID = Sensor(
+    id=SensorId(2),
+    name=_SENSOR.name,
+    manufacturer=_SENSOR.manufacturer,
+    model_name=_SENSOR.model_name,
+)
+
+_DIFFERENT_BOTH_ID = Sensor(
+    id=SensorId(2),
+    name=_SENSOR.name,
+    manufacturer=_SENSOR.manufacturer,
+    model_name=_SENSOR.model_name,
+)
```
is there a typo here?  They both seem to be the same.

llucax replied:
Ah, good catch. No, in the first version I included the microgrid ID too, as in 0.17 it is included, but to get it I needed to do an extra call to `GetMicrogridMetadata` just to get the microgrid ID. Since adding the microgrid ID in the next version is a backwards-compatible change I opted to leave the microgrid ID out, but forgot to remove the case for "both ID". Will remove.

shsms reviewed (commented):
Need to continue from `Add wrappers for sensor data`.

shsms commented on src/frequenz/client/microgrid/sensor.py:
```diff
@@ -48,3 +73,170 @@ def __str__(self) -> str:
+class SensorStateSample:
+    """A sample of state, warnings, and errors for a sensor at a specific time."""
+
+    sampled_at: datetime
```
wasn't it supposed to be `sample_time`?

llucax replied:
No, actually the current candidate is `snapshot_time` (and the message will be `SensorStateSnapshot`), but it is not changed yet, for stuff that it is already a bit unstable/changing, I just left the name used in 0.17. @tiyash-basu-frequenz do you think this is stable enough to adopt the new names in the Python API already?

llucax replied:
Thinking about this, there are **many** things that will change, even if they are considered fairly stable now, I think I prefer to leave the "modernization" of the client interface for a follow-up PR, let's not block this one on that as we are supposed to have people wanting to use this now.

llucax replied:
* https://github.com/frequenz-floss/frequenz-client-microgrid-python/issues/147

shsms commented on src/frequenz/client/microgrid/sensor.py:
```diff
@@ -48,3 +73,170 @@ def __str__(self) -> str:
+    UNSPECIFIED = 0
+    """Default value (this should not be normally used and usually indicates an issue)."""
+
+    TEMPERATURE = 1
```
Time to switch to pb2 names instead of numbers now?

llucax replied:
I'm still not convinced. This is a pure-python class/interface, all the dealing with protobuf conversion is done in the `_from_proto()` functions. I just happen to pick the same numeric IDs for convenience, to make conversion more easy.

What arguments do you have in favor of using the PB values here?

shsms replied:
> This is a pure-python class/interface

We are trying to build a client for the microgrid API.  Not a stand-alone thing.  This is just increasing the chances for bugs, especially for esoteric component types, which we might add support for only later.  To me, this is just overcomplicating things for no apparent benefit.

> What arguments do you have in favor of using the PB values here?

Single source of truth for enums without any manual work.
and just a few days back, I saw some API PRs changing the numbers for enums.

llucax replied:
I understand your point about this being very dependant on the defined protobuf stuff, but all wrappers are protobuf free. I know it is a very theoretical thing but I like it. :D

It is a very minor thing, but the alternative to me is also a very minor thing. The single point of truth is pointless once this is stable, is a write once, never ever touch it again, ever. So I'm not concerned at about about it.

That said, since it is controversial, and the API is currently NOT stable, I can see it might be problematic until we get there, and if we have some issue because of this, I don't want you in my face telling me I told you so, so I'll change it and leave the discussion for a distant future :P

shsms replied:
> I told you so

Oh, I really enjoy doing it, too bad.  But there will be other opportunities. :D

llucax committed:
Make `id` a public module

This module will be moved to `client-common` in the future, and there
will be more additions, so it makes sense that it is separated from the
microgrid module.

In an upcoming commit we'll also add a `BaseId` class to make defining
new IDs easier.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Add a base class for creating IDs

Defining new IDs via the new base class is much simpler and concise now,
and avoids a lot of code duplication.

Please see the module documentation for more details.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Use enum values from protobuf enums

We need to remove some `SensorErrorCode`s because they are not defined
in 0.15, they were included as a future-compatibility with 0.17.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax commented:
Added some new and some fixup! commits to address the comments. One minor advantage of using `int`s is we can add values that are not present in the protobuf yet (but OK, we could still do that only for the non-existing values, but I decided to remove them for now).

shsms commented:
Will check in the morning.

shsms commented on src/frequenz/client/microgrid/id.py:
```diff
@@ -1,157 +1,218 @@
 # Copyright © 2025 Frequenz Energy-as-a-Service GmbH
 
-"""Strongly typed IDs for microgrids, components and sensors."""
+r'''Provides strongly-typed unique identifiers for entities.
```
This is very nice, also useful outside the microgrid like for the GridpoolID, etc. and also the reporting client.  Maybe this /and/ the subclasses should go to the base client.

llucax replied:
Yes, next step is move lots of stuff in here to `client-common` and `client-base`.
